### PR TITLE
[InstCombine] Support and/or in `getFreelyInvertedImpl` using DeMorgan's Law

### PR DESCRIPTION
**Pull request by dtcxzyw**
This patch adds the support for and/or in `getFreelyInvertedImpl` using DeMorgan's Law:
```
(~(A | B)) -> (~A & ~B)
(~(A & B)) -> (~A | ~B)
```
Alive2: https://alive2.llvm.org/ce/z/Uig8-j

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp [hunk 1]
```diff
@@ -2596,6 +2596,32 @@ Value *InstCombiner::getFreelyInvertedImpl(Value *V, bool WillInvertAllUses,
...
+    }
+
+    return nullptr;
+  };
+
+  if (match(V, m_Or(m_Value(A), m_Value(B))))
+    return TryInvertAndOrUsingDeMorgan(Instruction::And, A, B);
+
+  if (match(V, m_And(m_Value(A), m_Value(B))))
+    return TryInvertAndOrUsingDeMorgan(Instruction::Or, A, B);
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:2623]: Should also support logical and/or? (Note that the above select code does *not* handle them already.)

Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:2623]: Alive2: https://alive2.llvm.org/ce/z/Uig8-j

Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstructionCombining.cpp]: Done.

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp [hunk 2]
```diff
@@ -2596,6 +2596,45 @@ Value *InstCombiner::getFreelyInvertedImpl(Value *V, bool WillInvertAllUses,
...
+    return TryInvertAndOrUsingDeMorgan(Instruction::Or, /*IsLogical=*/false, A,
+                                       B);
+
+  if (match(V, m_LogicalOr(m_Value(A), m_Value(B))))
+    return TryInvertAndOrUsingDeMorgan(Instruction::And, /*IsLogical=*/true, A,
+                                       B);
+
+  if (match(V, m_LogicalAnd(m_Value(A), m_Value(B))))
+    return TryInvertAndOrUsingDeMorgan(Instruction::Or, /*IsLogical=*/true, A,
+                                       B);
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:2636]: I'd just keep the m_Logical cases and then do `bool IsLogical = isa<SelectInst>` inside the helper.

Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:2636]: `m_Logical` only matches `and/or` with boolean types.

https://github.com/llvm/llvm-project/blob/f2d02ce04fe679c7cc806722a3c4a303f970f95f/llvm/include/llvm/IR/PatternMatch.h#L2739-L2757

Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:2636]: > Note the select code intentionally skips logicaland/logicalor patterns via: `shouldAvoidAbsorbingNotIntoSelect`

Yeah, logical ors/ands will be handled here :)

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:2636]: > > Note the select code intentionally skips logicaland/logicalor patterns via: `shouldAvoidAbsorbingNotIntoSelect`
> 
> Yeah, logical ors/ands will be handled here :)

I had misread nikics comment, thought he meant just let the select logic handle this. Delete when I saw my err :)

**Comment by goldsteinn** [comment]
LGTM, wait on nikic

**Review by nikic — APPROVED** [review]
LGTM

**Comment by vitalybuka** [comment]
We probably need to revert this one before figuring out what to do about #86880

**Comment by thurstond** [comment]
> We probably need to revert this one before figuring out what to do about #86880

topperc mentioned in https://github.com/llvm/llvm-project/issues/86880#issuecomment-2024110702 that it "post-RA machine copy propagation removed a copy that was needed to make the code correct", so it's probably not worthwhile reverting this innocent patch that just happened to unluckily trigger the bug.

**Comment by topperc** [comment]
> > We probably need to revert this one before figuring out what to do about #86880
> 
> topperc mentioned in [#86880 (comment)](https://github.com/llvm/llvm-project/issues/86880#issuecomment-2024110702) that it "post-RA machine copy propagation removed a copy that was needed to make the code correct", so it's probably not worthwhile reverting this innocent patch that just happened to unluckily trigger the bug.

I'm debugging MCP now.

**Comment by thurstond** [comment]
Thank you @topperc! :-)

**Pull request merged**